onnxruntime/core/optimizer/expand_elimination.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/expand_elimination.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status ExpandElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  if (graph_utils::RemoveNode(graph, node)) {
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+  }
+
+  return Status::OK();
+}
+
+bool ExpandElimination::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger&) const {
+  // 1. Get input shape.
+  const auto* input_shape = node.InputDefs()[0]->Shape();
+  if (input_shape == nullptr) {
+    return false;
+  }
+
+  std::vector<int64_t> dim_values;
+  for (int i = 0; i < input_shape->dim_size(); i++) {
+    auto dim = input_shape->dim(i);
+    if (!utils::HasDimValue(dim)) {
```

Comment:
>HasDimValue [](start = 16, length = 11)

Do you want to handle the DimParam case? 

if input is [batch, seq, 512] and target shape is [1, 1, 512]

expand can still be removed